be/src/exec/pipeline/context_base.h

```diff
@@ -0,0 +1,63 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021 StarRocks Limited.
+
+#pragma once
+
+#include "common/status.h"
+#include "runtime/runtime_state.h"
+
+namespace starrocks::pipeline {
+
+// The complex ExecNode is usually decomposed to multiple operators,
+// which share some variables in a context, and have dependency relationships between them.
+// (For example, HashJoinNode is composed to HashJoinBuildOperator and HashJoinProbeOperator,
+// which share a HashJoiner context. And HashJoinProbeOperator depends on HashJoinBuildOperator.)
+// Assume OpA depends on OpB.
+// However, OpA can be finished earlier than OpB, if it's successor operator is finished early.
+// Therefore, there are two problems which needs to be considered:
+// 1. OpB should perceive that OpA has been finished early.
+// 2. Context object can be released only if both OpA and OpB are closed,
+//    because the close order of OpA and OpB are uncertain.
+class ContextBase {
+public:
+    ContextBase() = default;
+    ~ContextBase() = default;
+
+    ContextBase(const ContextBase&) = delete;
+    ContextBase(ContextBase&&) = delete;
+    ContextBase& operator=(ContextBase&&) = delete;
+    ContextBase& operator=(const ContextBase&) = delete;
+
+    // For pipeline, it is called by unref() when the last operator is unreffed.
+    // For non-pipeline, it is called by close() of the exec node directly
+    // without calling ref and unref.
+    virtual Status close(RuntimeState* state) = 0;
+
+    // ref_no_barrier and unref are used to close context
+    // when the last running related operator is closed.
+    // - ref_no_barrier is called by operator::constructor() at the preparation stage.
+    // - unref is called by operator::close() at the close stage.
+    // It is the guaranteed by the dispatcher queue that the increment operations
+    // by ref_no_barrier() are visible to unref(), so we needn't barrier here.
+    void ref_no_barrier() { _num_running_operators.fetch_add(1, std::memory_order_relaxed); }
```

Comment:
Why the name need `no_barrier`?

Reply:
The reason is that we only use **relaxed** memory order instead of more serious memory order in `ref_no_barrier()`.

Reply:
I see. but i think for method name, ref is enough.

Reply:
Sounds great. Let me rename it.